### PR TITLE
[Features] Support `input_context` arguments for `input_fn` in DistributionStrategy

### PR DESCRIPTION
**Pull request by wangsiyu**
The InputContext in DistributionStrategy is very useful for estimator training especially for reading data shards according to input_pipeline_id in multiple nodes. This PR does some works in compatibility for mode,params and config arguments in input_fn.
Please take a look. Thanks very much. @yuefengz

### CODE REVIEW

**Comment by yangjunpro** [comment]
@yuefengz 
Yuefeng,

Could you please take a look at this PR?

This PR has been pending for one week.

Thanks

File: tensorflow_estimator/python/estimator/estimator.py [hunk 1]
```diff
@@ -1036,15 +1036,51 @@ def _validate_features_in_predict_input(self, result):
                       'QueueRunner. That means predict yields forever. '
                       'This is probably a mistake.')
 
-  def _get_iterator_from_input_fn(self, input_fn, mode, distribution=None):
-    if distribution is not None:
+  def _has_input_context_arg(self, input_fn_args):
+    # Only support 4 arguments at most. They are `mode`, `config`, `params`
```
Comment (yuefengz) [inline tensorflow_estimator/python/estimator/estimator.py]: @karmel Is this fine if we limit args within ["mode", "config", "params",  "input_context"]?

File: tensorflow_estimator/python/estimator/estimator.py [hunk 2]
```diff
@@ -1036,15 +1036,51 @@ def _validate_features_in_predict_input(self, result):
...
+    # and `input_context` for input_fn. The additional arg will be taken
+    # as the `input_context`. It should be the first one and it is useful 
+    # for distributio strategy.
+    optional_args = tuple(['mode', 'params', 'config'])
+    difference = set(input_fn_args).difference(set(optional_args))
+    if len(difference) > 0:
+      assert len(difference) == 1, ('Only one more arg should be taken as '
```
Comment (yuefengz) [inline tensorflow_estimator/python/estimator/estimator.py]: We prefer `raise ValueError`.

File: tensorflow_estimator/python/estimator/estimator.py (hunk 1 above)
Comment (wangsiyu) [reply tensorflow_estimator/python/estimator/estimator.py]: Actually, it seems strange  in `Estimator`. Because `input_fn` should not have args in `Estimator`. The document doesn't mentioned `config`, `params` and `mode` in `input_fn` and it will take effect only with these spelling. Does the users know this features ?

File: tensorflow_estimator/python/estimator/estimator.py (hunk 2 above)
Comment (wangsiyu) [reply tensorflow_estimator/python/estimator/estimator.py]: Thanks for comments. Assertion has been replaced by ValueError.

**Comment by wangsiyu** [comment]
@yuefengz Any update for me？ This has pended for long time. Thanks

File: tensorflow_estimator/python/estimator/estimator.py [hunk 3]
```diff
@@ -1036,15 +1036,53 @@ def _validate_features_in_predict_input(self, result):
...
                       'This is probably a mistake.')
 
-  def _get_iterator_from_input_fn(self, input_fn, mode, distribution=None):
-    if distribution is not None:
+  def _has_input_context_arg(self, input_fn_args):
+    # Only support 4 arguments at most. They are `mode`, `config`, `params`
+    # and `input_context` for input_fn. The additional arg will be taken
+    # as the `input_context`. It should be the first one and it is useful 
+    # for distributio strategy.
+    optional_args = tuple(['mode', 'params', 'config'])
```
Comment (karmel) [inline tensorflow_estimator/python/estimator/estimator.py]: All of this input arg parsing should be in one place; shouldn't _wrap_optional_args... handle this?

File: tensorflow_estimator/python/estimator/estimator.py [hunk 4]
```diff
@@ -1036,15 +1036,53 @@ def _validate_features_in_predict_input(self, result):
...
+    difference = set(input_fn_args).difference(set(optional_args))
+    if len(difference) > 0:
+      if len(difference) != 1:
+        raise ValueError('Only one more arg should be taken as ``'
+          '`input_context` of input_fn except for `mode`, `params` and '
+          '`config`, but found %d.' % len(difference))
+      if input_fn_args.index(list(difference)[0]) != 0:
+        raise ValueError('`%s` of input_fn will be taken as `input_context` '
+          'for DistributionStrategy usage. It should be the 1st arg of '
+          '`input_fn`' % list(difference)[0])
```
Comment (karmel) [inline tensorflow_estimator/python/estimator/estimator.py]: IIUC, this is not a backwards compatible change; anyone who had extra kwargs will break, correct? Right now, we don't pass extra args through, but I don't think having extra args breaks anything.

File: tensorflow_estimator/python/estimator/estimator.py [hunk 5]
```diff
@@ -1036,15 +1036,53 @@ def _validate_features_in_predict_input(self, result):
...
+    # as the `input_context`. It should be the first one and it is useful 
+    # for distributio strategy.
+    optional_args = tuple(['mode', 'params', 'config'])
+    difference = set(input_fn_args).difference(set(optional_args))
+    if len(difference) > 0:
+      if len(difference) != 1:
+        raise ValueError('Only one more arg should be taken as ``'
+          '`input_context` of input_fn except for `mode`, `params` and '
+          '`config`, but found %d.' % len(difference))
+      if input_fn_args.index(list(difference)[0]) != 0:
```
Comment (karmel) [inline tensorflow_estimator/python/estimator/estimator.py]: The logic here is hard to follow; can you add comments/use more straightforward checks to clarify what exactly this represents?

File: tensorflow_estimator/python/estimator/estimator.py [hunk 6]
```diff
@@ -1036,15 +1036,53 @@ def _validate_features_in_predict_input(self, result):
...
+          'for DistributionStrategy usage. It should be the 1st arg of '
+          '`input_fn`' % list(difference)[0])
+      return True
+    else:
+      return False
+
+  def _get_distribute_iterator_and_hooks(self,
+                                         input_fn,
+                                         mode,
+                                         distribution=None):
```
Comment (karmel) [inline tensorflow_estimator/python/estimator/estimator.py]: Docstrings for all new methods would be helpful here.

File: tensorflow_estimator/python/estimator/estimator.py (hunk 1 above)
Comment (karmel) [reply tensorflow_estimator/python/estimator/estimator.py]: In general, I would feel much more comfortable to adding input_context= to the set of kwargs we pass through; why does input_context have to be a positional arg, and the first one? Can we instead call with some established keyword so we can check and explicitly catch?

File: tensorflow_estimator/python/estimator/estimator.py [hunk 7]
```diff
@@ -1036,15 +1036,53 @@ def _validate_features_in_predict_input(self, result):
...
+    input_fn_args = function_utils.fn_args(input_fn)
+    has_input_context = self._has_input_context_arg(input_fn_args)
+    if has_input_context:
+      kwargs = self._wrap_optional_args_for_input_fn(input_fn_args, mode)
+      iterator = distribution.make_input_fn_iterator(
+          lambda context: input_fn(context, **kwargs))
+      input_hooks = [estimator_util.MultiHostDatasetInitializerHook(
+          iterator.initialize())]
```
Comment (karmel) [inline tensorflow_estimator/python/estimator/estimator.py]: Can you explain why this requires a different hook here?

File: tensorflow_estimator/python/estimator/estimator.py [hunk 8]
```diff
@@ -1129,13 +1186,7 @@ def _call_input_fn(self, input_fn, mode):
...
-    kwargs = {}
-    if 'mode' in input_fn_args:
-      kwargs['mode'] = mode
-    if 'params' in input_fn_args:
-      kwargs['params'] = self.params
-    if 'config' in input_fn_args:
-      kwargs['config'] = self.config
+    kwargs = self._wrap_optional_args_for_input_fn(input_fn_args, mode)
     with ops.device('/cpu:0'):
       return input_fn(**kwargs)
```
Comment (karmel) [inline tensorflow_estimator/python/estimator/estimator.py]: There are no tests here. Please make sure that all of this functionality is unit tested, and integration tested for the distribution path.

**Comment by yuefengz** [comment]
Gentle ping on this PR. We'll need this one as well.

**Comment by wangsiyu** [comment]
@yuefengz OK. I am on vacation due to the Spring Festival. I will update this code in the comming two days. Thanks for your reply.

**Comment by wangsiyu** [comment]
Hi, @yuefengz . I noticed the `distributed_dataset` has been deprecated and removed from `estimator.py`. This has simplifies the integration with `input_context` of `DistributionStrategy`. My code has been refined. Please check it again. I have tested with `simple_estimator_example.py` in tensorflow repo. And I also create an [PR](https://github.com/tensorflow/tensorflow/pull/25664) on this example .

File: tensorflow_estimator/python/estimator/estimator.py [hunk 9]
```diff
@@ -1076,8 +1076,14 @@ def _call_input_fn(self, input_fn, mode):
       kwargs['params'] = self.params
     if 'config' in input_fn_args:
       kwargs['config'] = self.config
-    with ops.device('/cpu:0'):
+    if input_context and 'input_context' in input_fn_args:
+      logging.info('`input_context` has been detected in `input_fn`.'
+                   'Placement willi be taken over by DistributionStrategy.')
```
Comment (rchao) [inline tensorflow_estimator/python/estimator/estimator.py]:
```suggestion
                   'Placement will be taken over by DistributionStrategy.')
```

Comment (wangsiyu) [reply tensorflow_estimator/python/estimator/estimator.py]: Misspelling has been fixed. Thanks.

File: tensorflow_estimator/python/estimator/estimator.py [hunk 10]
```diff
@@ -1076,8 +1076,14 @@ def _call_input_fn(self, input_fn, mode):
       kwargs['params'] = self.params
     if 'config' in input_fn_args:
       kwargs['config'] = self.config
-    with ops.device('/cpu:0'):
+    if input_context and 'input_context' in input_fn_args:
+      logging.info('`input_context` has been detected in `input_fn`.'
```
Comment (yuefengz) [inline tensorflow_estimator/python/estimator/estimator.py]: Maybe logging 'the `input_fn` accepts an `input_context` which will be given by Distribution Strategy.'?

File: tensorflow_estimator/python/estimator/estimator.py [hunk 11]
```diff
@@ -1076,8 +1076,14 @@ def _call_input_fn(self, input_fn, mode):
...
+                   'Placement will be taken over by DistributionStrategy.')
+      kwargs['input_context'] = input_context
```
Comment (yuefengz) [inline tensorflow_estimator/python/estimator/estimator.py]: You probably want to add `with ops.device('/cpu:0'):` for this branch as well?

Comment (wangsiyu) [reply tensorflow_estimator/python/estimator/estimator.py]: Yes, you are right. I thought the device placement should be managed by `InputFunctionIterator` because the `ops.device` also be called inside that function. But that placement seems not always for `cpu`. It is reasonable to add it as well here.

Comment (wangsiyu) [reply tensorflow_estimator/python/estimator/estimator.py]: Yes, you are right. I thought the device placement should be managed by `InputFunctionIterator` because the `ops.device` also be called inside that function. But that placement seems not always for `cpu`. It is reasonable to add it as well here.

File: tensorflow_estimator/python/estimator/estimator.py (hunk 10 above)
Comment (wangsiyu) [reply tensorflow_estimator/python/estimator/estimator.py]: Things have been changed.

**Comment by wangsiyu** [comment]
@yuefengz The code have been updated. Please check it again. Thanks.

File: tensorflow_estimator/python/estimator/estimator.py (hunk 8 above)
Comment (wangsiyu) [reply tensorflow_estimator/python/estimator/estimator.py]: Thanks for comments. I have changed my coding style. The integration test for distribution path has been updated in `tensorflow` repo. https://github.com/tensorflow/tensorflow/pull/25664

**Comment by wangsiyu** [comment]
@karmel Hi, I have update my code. Please check it again. Thanks.

File: tensorflow_estimator/python/estimator/estimator.py (hunk 8 above)
Comment (karmel) [reply tensorflow_estimator/python/estimator/estimator.py]: This change looks fine to me now, thanks, but we should think through how the testing of this will have to go in-- presumably that test fails without this change? So we have to push this, then that. Can we at least add a unit test here locally that checks the logic of passing the input context through the input fn?

Comment (wangsiyu) [reply tensorflow_estimator/python/estimator/estimator.py]: OK. I understand what you mean. I will add one unit test to check the `DistributionStrategy` pass but not really calling that logic.

**Comment by wangsiyu** [comment]
@yuefengz @karmel I have add unit test. Please check it again. Thanks.

**Comment by yuefengz** [comment]
@rthadur Would you mind merging this PR? Thank you!
